### ISSUE TITLE
Low level stack overflows using Jedi in Chrome

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#821]
If I type into the pyodide console here: https://pyodide-cdn2.iodide.io/v0.15.0/full/console.html
`import jedi; import numpy; jedi.Interpreter("numpy.split().", [globals()]).completions()`
the result is "Range Error: Maximum Call Stack Exceeded". The pyodide instance is not completely borked by this but it is pretty screwed up. For instance, there is an error left in `sys.exc_info()`:
```python
>>> import sys; sys.exc_info()
<class 'KeyError'>,KeyError(((), frozenset())),<traceback object at 0x4976ab0>
```
Trying to set a very small recursion limit will tell you that a 334 calls worth of glitched junk is left over on the Python stack:
```python
>>> sys.setrecursionlimit(10)
<... many lines of stack trace omitted>
RecursionError: cannot set the recursion limit to 10 at the recursion depth 334: the limit is too low
```
In a fresh repl this would say `cannot set the recursion limit to 10 at the recursion depth at the recursion depth 9`.

If I start out with `sys.setrecursionlimit(100)` before doing the Jedi call, then it will return a proper Python RecursionError. The initial value of `sys.getrecursionlimit()` is 315.

I did all of this with Chrome v. 86.0.4240.198 under Windows 10.

**Comment by hoodmane** [comment]
Note that the same code works completely fine in Firefox.
Even with a sufficiently low `recursionlimit` so that no low level stack overflows occur, there is still the problem that Jedi seems to only work in Chrome with extremely simple cases because in all other cases there is a stack overflow.

I saw somewhere mention of enlarging the Emscripten [TOTAL_STACK](https://github.com/emscripten-core/emscripten/blob/49029d44cbef0f535fcd48445300808c7efbdb10/src/settings.js#L113) constant, does that have any relationship with this issue? It seems like it might be an issue to raise with the chromium team, but I don't know if there is any chance of convincing them to enlarge the stack.

**Comment by rth** [comment]
Thanks for the report @hoodmane ! Interesting I can't reproduce with Chrome 80.0.3987 and Windows 10, it works fine there with 0.15.0.

Could you please try with the latest development version https://pyodide-cdn2.iodide.io/dev/full/console.html  (which also has a later version of jedi) ?

> enlarging the Emscripten TOTAL_STACK constant, does that have any relationship with this issue?

Yes, it could be worth a try, related to https://github.com/iodide-project/pyodide/issues/716

**Comment by hoodmane** [comment]
I cannot reproduce the error on https://pyodide-cdn2.iodide.io/dev/full/console.html because there seems to be some other problem with the Jedi configuration, it returns:

```python
  File "/lib/python3.8/site-packages/jedi/parser_utils.py", line 280, in get_cached_code_lines
    return parser_cache[grammar._hashed][path].lines
KeyError: '/lib/python3.8/site-packages/numpy/__init__.py'
```
Similar `KeyError`s happen with other imports (I can also trigger the error with commands from several other packages like `Pathlib`. The problem with the recursion limit being set too high is still present, which I can see by a more direct approah:

```python
>>> def f(): return 1 + f()
>>> f()
RangeError: Maximum call stack size exceeded
>>> sys.setrecursionlimit(10)
Error: Traceback (most recent call last):
  File "/lib/python3.8/site-packages/pyodide.py", line 68, in eval_code
    return eval(compile(expr, "<eval>", mode="eval"), ns, ns)
  File "<eval>", line 1, in <module>
RecursionError: cannot set the recursion limit to 10 at the recursion depth 369: the limit is too low
```
Again with Chrome v. 86.0.4240.198 under Windows 10.

This second test shows the same problem with Firefox 83.0 (also windows 10):
``python
>>> def f(): return 1 + f()
>>> f()
InternalError: too much recursion
>>> import sys
>>> sys.setrecursionlimit(10)
Error: Traceback (most recent call last):
  File "/lib/python3.8/site-packages/pyodide.py", line 68, in eval_code
    return eval(compile(expr, "<eval>", mode="eval"), ns, ns)
  File "<eval>", line 1, in <module>
RecursionError: cannot set the recursion limit to 10 at the recursion depth 431: the limit is too low
```

There seem to be two separate issues here:

1. The recursion limit is not set low enough to prevent low level errors when too much recursion occurs.
2. The low level stack size available in Chrome isn't really big enough to allow Jedi to do any useful work, with even the simplest Jedi queries nearly hitting the limit.

I have run into similar problems with Jedi 17, which I've been using for a project with Jedi (project is joint with @dalcde, he's been trying to upstream some of our changes).

**Comment by hoodmane** [comment]
I'm intrigued by PR #602, I wonder if they have some setup that prevents these problems or if they've only tried it out on such simple examples that they don't see the recursion errors. I would think these stack overflows would be a blocking issue for that.

**Comment by dalcde** [comment]
The current dev version of pyodide uses Jedi 1.7

**Comment by dalcde** [comment]
I can reproduce on linux with chromium 87.0.4280.88, but only on git master and not v0.15.0

**Comment by hoodmane** [comment]
I posted a question about this on Emscripten:
emscripten-core/emscripten#13043

**Comment by hoodmane** [comment]
And I opened an issue in V8 here:
https://bugs.chromium.org/p/v8/issues/detail?id=11269

**Comment by hoodmane** [comment]
Further reading shows that this is an issue of the value of the `--stack-size` v8 compilation flag used when compiling Chromium so I opened roughly the same bug report on Chromium:

https://bugs.chromium.org/p/chromium/issues/detail?id=1159217

**Comment by hoodmane** [comment]
I found some further explanation about the v8 `--stack-size` flag in the nodejs issues pages:

> `--stack_size` controls what V8 _thinks_ is the stack size, not the actual stack size. It's for overriding the default when the default doesn't match reality. V8's estimate is fairly conservative, there is normally no need to touch it.
> 
> If you set it higher than the actual limit, the process gets killed (on UNIX) by a SIGBUS signal when it exceeds the limit.

https://github.com/nodejs/node/issues/11091#issuecomment-281117576

**Comment by oeway** [comment]
I might hit a similar error here that only happens when I start to import scikit-image and using a custom asyncio loop that based on recursion.

Here is the error I got:
```
pyodide.asm.js:8 Uncaught (in promise) Error: Traceback (most recent call last):
  File "/lib/python3.8/site-packages/imjoy_rpc/connection/pyodide_connection.py", line 62, in <lambda>
    self.timeout_promise(10).then(lambda x: self.run_forever())
RecursionError: maximum recursion depth exceeded

    at _hiwire_throw_error (pyodide.asm.js:8)
    at _pythonexc2js (cdn.jsdelivr.net/pyodide/dev/full/pyodide.asm.wasm:wasm-function[372]:0x10793d)
    at __pyproxy_apply (cdn.jsdelivr.net/pyodide/dev/full/pyodide.asm.wasm:wasm-function[370]:0x107636)
    at Module.__pyproxy_apply (pyodide.asm.js:8)
    at Object.apply (pyodide.asm.js:8)
```

This error comes from the custom asyncio loop I made([see this issue](https://github.com/iodide-project/pyodide/issues/245#issuecomment-744072568)). It basically wraps the `setTimeout` function with a promise in js and use that to drive an infinite loop via recursion:
```
def run_forever(self):
        self._running = True
        try:
            self._do_tasks()
            self.timeout_promise(10).then(lambda x: self.run_forever())
        except Exception as exp:
            self._running = False
            raise exp
```
Full code [here](https://gist.github.com/oeway/9a2d5ca5b5920c2929cb05c9ff78e43d#file-eventsimulatorpyodide-py).

The loop runs normally without any issue, and the wired thing is that the RecursionError only occurs if I run `import skimage`.

**Comment by hoodmane** [comment]
Interestingly this no longer reproduces on `/dev/full/` but it was not fixed by #1102 it also fails on 0.17.0a2. No clue what the deal with that is, but probably the problem is still present we just need a slightly more complicated trigger.

@rth These tagged alphas are really great, we should make more of them in the future.

**Comment by hoodmane** [comment]
It really seems to work now!! What the heck!

**Comment by leopsidom** [comment]
Found another command that still triggers the error:

```
import jedi; import numpy; jedi.Interpreter("numpy.in1d([1,2,3],[2]).", [globals()]).complete()
```

> Uncaught RangeError: Maximum call stack size exceeded

pyodide version: 0.17.0

**Comment by hoodmane** [comment]
Great thanks! I can reproduce this.

**Comment by hoodmane** [comment]
Okay this is closed by #2019.

**Comment by leopsidom** [comment]
Thank you @hoodmane. It's amazing this actually gets fixed. And I'm wondering if there could be other edge cases. Can't wait to test it out with the next version of pyodide.
